### PR TITLE
Role subcollection

### PR DESCRIPTION
**Pull request by phdodo**

### CODE REVIEW

File: firebase.json [hunk 1]
```diff
@@ -1,7 +1,7 @@
 {
   "firestore": {
     "predeploy": "npm run database:predeploy",
-    "rules": "firestore.min.rules",
```
Comment (brianneisler) [inline firebase.json:4]: Should probably leave this as min for now

File: firestore.rules [hunk 2]
```diff
@@ -1965,6 +1966,60 @@ service cloud.firestore {
...
+
+    function validateRole(resource) {
+      return isString(resource.data.roleName) &&
+      resource.id == resource.data.roleName &&
+      userExistsBefore(resource.data.userId) &&
+      (resource.data.roleName == 'admin' || resource.data.roleName == 'hil')
+    }
+
+    function checkReadRole(auth, resource) {
+      return currentUserCanReadPublicEntity(auth, resource, false)
```
Comment (brianneisler) [inline firestore.rules:1996]: These entities should probably be private read. Don't want a user to be able to scan for admin account ids.

File: src/constants/index.js [hunk 3]
```diff
@@ -13,6 +13,7 @@ import * as OSType from './OSType'
 import * as Overlay from './Overlay'
 import * as PageContentFormat from './PageContentFormat'
 import * as PlatformType from './PlatformType'
+import * as Role from './Role'
```
Comment (brianneisler) [inline src/constants/index.js:16]: Don't forget to export this at the bottom

File: src/modules/role/schemas/Role.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+const Role = {
+  collectionName: 'Roles',
+  idField: 'roleName',
+  name: 'Role',
+  parentRefIdField: 'userId',
+  parentSchema: UserRole,
+  schema: Entity.schema.keys({
+    userId: Id.schema.required(),
+    roleName: String.schema.required()
```
Comment (brianneisler) [inline src/modules/role/schemas/Role.js:13]: Can limit this with the allow statement to only be the enumerable values of Role

**Review by brianneisler — CHANGES_REQUESTED** [review]
Just a few tweaks to be made. Also, note that in the firebase.rules you will also need to adjust the function that checks whether the user is an admin.

File: src/constants/index.js (hunk 3 above)
Comment (phdodo) [reply src/constants/index.js:16]: moved it to the constants folder in role module

File: src/modules/role/schemas/Role.js (hunk 4 above)
Comment (phdodo) [reply src/modules/role/schemas/Role.js:13]: I imported them but then didn't use them 😅

File: firestore.rules [hunk 5]
```diff
@@ -2036,6 +2042,60 @@ service cloud.firestore {
...
+    }
+
+    function checkCreateRole(auth, id, index) {
+      let role = getRoleAfter(id, index);
+      return role != null && validateRole(role) &&
+        currentUserIsServiceAccount(auth)
+    }
+
+    function checkUpdateRole(auth, id, index) {
+      let queueEntity = getRoleAfter(id, index);
```
Comment (brianneisler) [inline firestore.rules:2082]: think this was meant to be `let role = ...`

**Review by brianneisler — APPROVED** [review]
Looks great! thanks @phdodo!

File: firestore.rules (hunk 5 above)
Comment (phdodo) [reply firestore.rules:2082]: of course 😅

**Pull request merged**